### ISSUE TITLE
Shouldn't @Value.Default imply @Value.Parameter(false)?

### ISSUE THREAD

**Issue by Stephan202** [immutables/immutables#524]
We generate classes using `@Value.Style(allParameters = true)`. Some classes have non-required attributes with default values. The associated methods are thus annotated with both `@Value.Default` and `@Value.Parameter(false)`. But it seems to me the latter annotation is redundant.

So my question is: can the definition of `@Value.Default` be updated to imply `@Value.Parameter(false)`, such that the latter annotation can be omitted? (Or are there scenarios in which `@Value.Default` and `@Value.Parameter(true)` make sense...?)

**Comment by elucash** [comment]
One one hand, that stuff is independent. Put `@Value.Parameter` and it will turn attribute into constructor parameter, but the builder will still have it as non-mandatory, defaulted value. But on the other hand, I do find it somewhat attractive to treat default attributes as non-parameters `@Value.Parameter(false)` when `@Value.Style(allParameters = true)` is enabled. The only things I worry about is if it will break people's code. Do you see any possible scenarios where the current behavior makes more sense? I can think of a bulk construction of objects avoiding builder allocations. In this case, it might be useful to initialize all the attributes handling default values behind the scenes. But I'm not sure it makes any sense.

**Comment by Stephan202** [comment]
> I can think of a bulk construction of objects avoiding builder allocations. In this case, it might be useful to initialize all the attributes handling default values behind the scenes. But I'm not sure it makes any sense.

I'm afraid I don't quite understand that scenario. Any chance you can show that in code? :)

> Do you see any possible scenarios where the current behavior makes more sense?

I have yet to come up with an example myself. (Hence the ticket ;).)

**Comment by elucash** [comment]
BTW #439 seems to be about the same

**Comment by Stephan202** [comment]
You're right, that ticket boils down to almost the same, except that I'm not interested in the signature of the internal constructor; just the signature of the public factory method.

(I did search before filing a ticket, but either didn't find it or failed to recognize it as a match. My apologies.)

**Comment by elucash** [comment]
It was not an intention to dissmiss something as dulicate, but rather find som common rationale. I do like the idea, just thinking of a ways how to not break stuff )

**Comment by elucash** [comment]
while thinking about it, the question arises, what to do with collections, optionals and other special types along with default attributes. My impression is that `allParameters` as it is implemented now was primarily suited for sort of integration with serialization/deserialization solutions to have low-allocation way to fully construct objects. It's expected that all optional, collections and defaultable attributes can be set at will in a single pass.
Maybe we need something like separate `allMandatoryParameters` setting to only enable turning mandatory attributes into constructor parameters.

**Comment by Stephan202** [comment]
> My impression is that `allParameters` as it is implemented now was primarily suited for sort of integration with serialization/deserialization solutions to have low-allocation way to fully construct objects.

Interesting. We're actually using it for (almost) all our classes i.c.w. "regular" invocations; the code base from which we're migrating has similar (manually written) factory methods.

>  Maybe we need something like separate `allMandatoryParameters` setting to only enable turning mandatory attributes into constructor parameters.

If "mandatory" means "without explicit default", then that would work for us. If this setting would also exclude all `Optional`s, that might still work or even be a good idea (would have to check...), as long as they can be added using `@Value.Parameter(true)` when necessary. I'm not so sure about excluding collection attributes. I can see that an empty collection would be the (implicit) default, but at least in our code base we're almost invariably providing non-empty collections to the factory methods.

Crazy idea (I haven't thought this through yet; not even sure I like it myself; just dumping an idea here):

```java
// Note that these types overlap; they do not neatly partition the type space.
enum ParameterType {
    ANY, // Same as `allParameters = true`
    WITH_DEFAULT,
    WITHOUT_DEFAULT,
    OPTIONAL,
    COLLECTION // Includes (multi)maps?
}
@Value.Style(
    ...
    parameters = ParameterType[]
    ...
)
```

**Comment by scranen** [comment]
> I do find it somewhat attractive to treat default attributes as non-parameters

From the deserialisation point of view I think this approach would require you to then _also_ supply a factory method with _all_ fields as parameters, otherwise you would always have to resort to using a builder for deserialisation of classes with optional fields. I think there are two usage scenarios:

- Code optimised for readability / maintainability: you don't want to supply absent optional parameters and you want to be protected from making mistakes. In that case you'd like to be steered towards using a builder for classes with optional fields, where the builder can have a factory method that takes all mandatory fields as parameters (so `ImmutableFoo.builder(mandatory1, mandatory2).optional1(optional1).build()`), and perhaps even a static factory method for all mandatory parameters itself in case this is an often-used scenario (so `ImmutableFoo.of(mandatory1, mandatory2)`).
- Code optimised for speed or for code generation: you want to always use a static factory method that includes all parameters, so you can use `ImmutableFoo.of(mandatory1, mandatory2, optional1)`. This is a less nice interface because it requires you to give values for absent parameters, but does allow for efficient creation of objects and simple interfacing with serialisation/deserialisation libraries.

Given that many users will likely be working in the context of a serialisation/deserialisation engine, you will usually need to cater for the second scenario. I could imagine that the factory method that includes optional parameters could have a configurable name so developers can be steered away from using it in hand-written code (call it `ImmutableFoo.jacksonCreate(...)` for instance if you're using Jackson). To customise the API generated for manual use, I think the `ParameterType` suggested by @Stephan202 gives more flexibility for future feature requests.

**Comment by elucash** [comment]
@scranen yes, that was my impression. It's just that we cannot redesign `allParameters`, quite a lot people are already using it. @Stephan202 as much as I like the flexible approach, I feel really few people would ever use it. You know, it feels like the right way to do it, but overkill at the same time. Suggesting to pick up some fine-enough preconfigured solution. Immutables already have some amount of features I would regret of: they are flexible but unnecessary. How about picking the best inclusion formula for `allMandatoryParameters`?

**Comment by Stephan202** [comment]
Alright @elucash; in that case I suggest for `allMandatoryParameters` to leave out `@Value.Default` and `Optional` parameters. But before we make a final decision, let me migrate some more of our code base to get a better feeling for this. Will get back to you.

**Comment by elucash** [comment]
`allMandatoryParameters` released in 2.4.4. thank you!
